packages/agents/src/react.tsx

```diff
@@ -27,18 +27,126 @@ function camelCaseToKebabCase(str: string): string {
   return kebabified.replace(/_/g, "-").replace(/-$/, "");
 }
 
+type QueryObject = Record<string, string | number | boolean | null | undefined>;
+
+const queryCache = new Map<
+  unknown[],
+  {
+    promise: Promise<QueryObject>;
+    refCount: number;
+    expiresAt: number;
+    cacheTtl?: number;
+  }
+>();
+
+function deepEqual(a: unknown, b: unknown): boolean {
+  if (a === b) return true;
+  if (a === null || b === null || a === undefined || b === undefined)
+    return a === b;
+
+  if (typeof a === "object" && typeof b === "object") {
+    const aObj = a as Record<string, unknown>;
+    const bObj = b as Record<string, unknown>;
+    const aKeys = Object.keys(aObj);
+    const bKeys = Object.keys(bObj);
+
+    if (aKeys.length !== bKeys.length) return false;
+
+    for (const key of aKeys) {
+      if (!bKeys.includes(key) || !deepEqual(aObj[key], bObj[key])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return false;
+}
+
+function arraysEqual(a: unknown[], b: unknown[]): boolean {
+  if (a.length !== b.length) return false;
+
+  for (let i = 0; i < a.length; i++) {
+    if (!deepEqual(a[i], b[i])) return false;
```

Comment:
this should actually be a shallow equal, not a deep equal (following the same rules as react deps)